core/platform/email/mailgun_email_services.py

```diff
@@ -0,0 +1,122 @@
+import requests
+
+from core import counters
+import feconf
+
+from google.appengine.api import app_identity
+
+MAILGUN_PUBLIC_DOMAIN_URL = "https://api.mailgun.net/v3/address/validate"
+MAILGUN_PUBLIC_API_KEY = "pubkey-5ogiflzbnjrljiky49qxsiozqef5jxp7"
+def is_email_valid(email):
+    """ Check whether recipient and sender email address are valid or not.
+    Mailgun uses public api key and public domain url to do this check. """
+    response = (
+        requests.get(
+            MAILGUN_PUBLIC_DOMAIN_URL, auth=("api", MAILGUN_PUBLIC_API_KEY),
+            params={"address": email}))
+    return response.json()['is_valid']
+
+def send_mail(
+        sender_email, recipient_email, subject, plaintext_body,
+        html_body, bcc_admin=False):
+    """Sends an email using mailgun api.
+
+    In general this function should only be called from
+    email_manager._send_email().
+
+    Args:
+      - sender_email: str. the email address of the sender. This should be in
+          the form 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'.
+      - recipient_email: str. the email address of the recipient.
+      - subject: str. The subject line of the email.
+      - plaintext_body: str. The plaintext body of the email.
+      - html_body: str. The HTML body of the email. Must fit in a datastore
+          entity.
+      - bcc_admin: bool. Whether to bcc feconf.ADMIN_EMAIL_ADDRESS on the email.
+
+    Raises:
+      Exception: if the configuration in feconf.py forbids emails from being
+        sent.
+      ValueError: if 'sender_email' or 'recipient_email' is invalid, according
+        to App Engine.
+      Exception: if mailgun api key is not stored in feconf.MAILGUN_API_KEY.
+      Exception: if mailgun domain name is not stored in
+        feconf.MAILGUN_DOMAIN_NAME.
+      (and possibly other exceptions, due to mail.send_mail() failures)
+    """
+    if not feconf.MAILGUN_API_KEY:
+        raise Exception('Mailgun API key is not available.')
+
+    if not feconf.MAILGUN_DOMAIN_NAME:
+        raise Exception('Mailgun domain name is not set.')
+
+    mailgun_domain_name = (
+        'https://api.mailgun.net/v3/%s/messages' % feconf.MAILGUN_DOMAIN_NAME)
+
+    if recipient_email == feconf.ADMIN_EMAIL_ADDRESS:
+        if feconf.CAN_SEND_EMAILS_TO_ADMIN:
+            if not is_email_valid(recipient_email):
+                raise ValueError(
+                    'Malformed admin email address: %s' % recipient_email)
+
+            app_id = app_identity.get_application_id()
+            body = '(Sent from %s)\n\n%s' % (app_id, plaintext_body)
+
+            requests.post(
+                mailgun_domain_name, auth=('api', feconf.MAILGUN_API_KEY),
+                data={
+                    'from': sender_email,
+                    'to': recipient_email,
+                    'subject': subject,
+                    'text': body})
+            counters.EMAILS_SENT.inc()
+        return
+
+    if not feconf.CAN_SEND_EMAILS_TO_USERS:
```

Comment:
You know, I think it's better to try and refactor this whole method (and the one in gae_email_services) to be simpler., and not have a special case for the admin. Basically something along the lines of:

  if not CAN_SEND_EMAILS: return
  if not is_email_valid: return
  requests.post(...)

Historically, the can_send_emails_to_admin() function was written first, long before the general one -- that's why it's still there. But from a holistic design perspective there shouldn't really be that special case, and this layer of the app shouldn't care whether the recipient is an admin or not. So, I also suggest that we combine CAN_SEND_EMAILS_TO_USERS and CAN_SEND_EMAILS_TO_ADMIN into just one constant: CAN_SEND_EMAILS -- it's about time we tidied that up, too!